### PR TITLE
Second deliverable

### PR DESCRIPTION
**Pull request by orlandorode97**
Deliverable details: https://classroom.google.com/c/Mzk0MDUzNTg0MjQz/a/Mzk0MDUzNTg0Mjc2/details

### Changes
- Refactored Urban Dictionary structs
- Implemented error handling
- Added unit testing for:
   - Error Handling
   - Urban dictionary API
   - Middleware
   - Definition Controller
   - Definition Presenter
   - Definition Repository
   - Definition Interactor
 - Refactor endpoint resources and controllers

### CODE REVIEW

File: domain/model/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+func TestEncodeError(t *testing.T) {
+	tests := []struct {
+		name        string
+		err         error
+		code        int
+		contentType string
+	}{
+		{"not found error type", ErrNotFound{}, http.StatusNotFound, "application/json; charset=utf-8"},
```
Comment (StradivariT) [inline domain/model/errors_test.go]: When defining your test cases, please define each field explicitly, it helps with readability, for example:

```
{
  name: "not found error type", 
  err: ErrNotFound{}, 
  code: http.StatusNotFound, 
  contentType: "application/json; charset=utf-8",
}
```

Even if it makes the file bigger, no problem

File: domain/model/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		{"not found in CSV error type", ErrNotFoundInCSV{}, http.StatusNotFound, "application/json; charset=utf-8"},
+		{"missing api key error type", ErrMissingApiKey{}, http.StatusForbidden, "application/json; charset=utf-8"},
+		{"missing field error type", ErrMissingField{}, http.StatusBadRequest, "application/json; charset=utf-8"},
+		{"invalid data error type", ErrInvalidData{}, http.StatusBadRequest, "application/json; charset=utf-8"},
+		{"random error type", errors.New("random unknown error"), http.StatusInternalServerError, "application/json; charset=utf-8"},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			w := httptest.NewRecorder()
```
Comment (StradivariT) [inline domain/model/errors_test.go:58]: I don't think anything from lines 31 - 35 are needed, right? I think you only need the ResponseWriter

File: infrastructure/api/api.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+		line, err := csvReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			fmt.Println(err.Error())
+			continue
+		}
+		if id == line[0] {
+			idCsv, _ := strconv.Atoi(line[0])
```
Comment (StradivariT) [inline infrastructure/api/api.go]: Don't ignore this error, handle it

File: infrastructure/api/api.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	for _, definition := range definitionsList.Definitions {
+		id := strconv.Itoa(int(definition.Defid))
+		err = csvWriter.Write([]string{
+			id,
+			definition.Word,
+			definition.WrittenOn,
+			definition.Definition,
+			definition.Permalink,
+			definition.Example})
+		if err != nil {
```
Comment (StradivariT) [inline infrastructure/api/api.go:172]: You should log the error and probably also return here

File: infrastructure/api/api.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+func (u *UrbanDictionary) UpdateCSVPath(path string) {
+	u.CSVPath = path
+}
+
+// GetDefinitions reach out urban dictionary API by term
+func (u *UrbanDictionary) GetDefinitions(term string) (*model.List, error) {
+	var list *model.List
+	url := fmt.Sprintf("%s?term=%s", u.ApiURL, term)
+	request, err := http.NewRequest(http.MethodGet, url, nil)
```
Comment (StradivariT) [inline infrastructure/api/api.go:57]: I suggest reading this article: https://www.thegreatcodeadventure.com/mocking-http-requests-in-golang/

It shows how you can reestructure a bit your code to be able to mock http calls, when doing tests you don't want to call the actual server

File: infrastructure/api/api_test.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		urbanClient := NewUrbanDictionary(testApiKey)
+		urbanClient.UpdateCSVPath("/usr/local/definitions.csv")
+		assert.Equal(t, urbanClient.CSVPath, "/usr/local/definitions.csv")
+	})
+}
+
+func TestGetDefinitions(t *testing.T) {
+	urbanClint := NewUrbanDictionary(testApiKey)
+	t.Run("success - definition by term", func(t *testing.T) {
+		definitions, err := urbanClint.GetDefinitions("ha")
```
Comment (StradivariT) [inline infrastructure/api/api_test.go]: I suggest taking a look at the article I shared previously to mock the http calls, you don't want to call the actual http server on tests. You'll need to refactor a bit your api service to be able to mock the http client

File: infrastructure/api/api_test.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+func TestGetDefinitionById(t *testing.T) {
+	urbanClint := NewUrbanDictionary(testApiKey)
+	t.Run("success - definition found in CSV", func(t *testing.T) {
+		urbanClint.UpdateCSVPath("../../data/definitions_test.csv")
+		definitions, err := urbanClint.GetDefinitionById("10593002")
+		assert.Nil(t, err)
+		assert.NotNil(t, definitions)
+	})
+
+	t.Run("success - definition not found in CSV", func(t *testing.T) {
```
Comment (StradivariT) [inline infrastructure/api/api_test.go]: If you expect an error from your function call, then the case is not a success, it should be a fail

File: infrastructure/api/api_test.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		urbanClint.UpdateCSVPath("../../data/definitions_test.csv")
+		definitions, err := urbanClint.GetDefinitionById("1232")
+		assert.EqualValues(t, model.ErrNotFoundInCSV{"1232"}, err)
+		assert.Nil(t, definitions)
+	})
+}
+
+func TestOpen(t *testing.T) {
+	urbanClint := NewUrbanDictionary(testApiKey)
```
Comment (StradivariT) [inline infrastructure/api/api_test.go]: Minor typo `urbanClint` -> `urbanClient`. Check all over your file for this

File: infrastructure/api/api_test.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	t.Run("success - open CSV file", func(t *testing.T) {
```
Comment (StradivariT) [inline infrastructure/api/api_test.go]: You should also add a test to see what happens when you provide a wrong path

File: infrastructure/api/api_test.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+)
+
+var (
+	testApiKey = "cc2464e4a1mshb5ceeca91e5a6adp1fa80bjsn4b48e2408b87"
+)
+
+func TestUpdateCSVPath(t *testing.T) {
+	t.Run("set valid CSV path", func(t *testing.T) {
+		urbanClient := NewUrbanDictionary(testApiKey)
+		urbanClient.UpdateCSVPath("/usr/local/definitions.csv")
```
Comment (StradivariT) [inline infrastructure/api/api_test.go]: I don't think you should use absolute paths, what if someone else downloads the repo and tries to run the tests? This could work in your machine but not in others, the csv should be on the repo

File: infrastructure/api/api_test.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+	t.Run("success - open CSV file", func(t *testing.T) {
+		urbanClint.UpdateCSVPath("../../data/definitions_test.csv")
+		_, err := urbanClint.Open()
+		assert.Nil(t, err)
+	})
+}
+
+func TestRead(t *testing.T) {
+	urbanClint := NewUrbanDictionary(testApiKey)
+	t.Run("success - reading content of the CSV", func(t *testing.T) {
```
Comment (StradivariT) [inline infrastructure/api/api_test.go]: Add a test case for a failure scenario, also in the test below

File: main.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+	"net/http"
+	"os"
+
+	"github.com/OrlandoRomo/academy-go-q32021/infrastructure/api"
+	"github.com/OrlandoRomo/academy-go-q32021/infrastructure/router"
+	"github.com/OrlandoRomo/academy-go-q32021/registry"
+)
+
+func main() {
+	var UrbanDictionaryApiKey = flag.String("urban-dictionary-api-key", envString("URBAN_DICTIONARY_API_KEY", "55a4103f9amsh99bbd454b45f012p180fb6jsnf682ba32da43"), "Urban Dictionary api key")
```
Comment (StradivariT) [inline main.go:15]: Don't insert sensitive data as api-keys into the repo, you could probably get this info from a config file

File: registry/registry_test.go [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+		{
+			name:        "failure - urban client nil",
+			message:     "registry app with a nil urban client",
+			urbanClient: nil,
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			if test.urbanClient == nil {
```
Comment (StradivariT) [inline registry/registry_test.go]: It seems like this if statements are redundant, why double checking if something is nil or not? Assert already takes care of that, it would be better to add a field to the tests to indicate if you expect a nil or not, and do something like `assert.Equal(t, test.expectedClient, test.urbanClient)`

And call the NewUrbanDictionary function inside this test body

File: registry/registry_test.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+			}
+			if test.urbanClient != nil {
+				assert.NotNil(t, test.urbanClient, test.message)
+			}
+		})
+	}
+}
+
+func TestNewAppController(t *testing.T) {
+	appController := NewRegistry(nil).NewAppController()
```
Comment (StradivariT) [inline registry/registry_test.go:44]: I find it a bit odd to use the `for` template for some tests, and then this individual `t.Run(...)` format on different tests, please be consistent, I would recommend using always the `for` approach

File: registry/registry_test.go [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+	t.Run("success - return AppController type", func(t *testing.T) {
+		assert.IsType(t, controller.AppController{}, appController, "app contoller type received")
+	})
+
+	t.Run("failure - return another type", func(t *testing.T) {
+		assert.NotSame(t, controller.AppController{}, struct{}{}, "expected AppControll type")
```
Comment (StradivariT) [inline registry/registry_test.go]: Your message seems incomplete `expected AppControll type`

File: usercase/interactor/definition_interactor_test.go [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			mockRepo := MockDefinitionsRepo{}
+			mockPresenter := MockDefinitionPresenter{}
+			mockRepo.On("GetDefinitionById", test.definitionId).Return(test.response, test.error)
+
+			interactor := NewDefinitionInteractor(mockRepo, mockPresenter)
+			definitions, _ := interactor.urbanDictionaryRepository.GetDefinitionById(test.definitionId)
+			id, _ := strconv.Atoi(test.definitionId)
```
Comment (StradivariT) [inline usercase/interactor/definition_interactor_test.go]: Why not just making this field an int instead of doing this conversion?

File: usercase/interactor/definition_interactor_test.go [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+			if definitions.Definitions[0].Defid == id {
```
Comment (StradivariT) [inline usercase/interactor/definition_interactor_test.go]: Why this redundant if statements? The assert.Equal will take care of validating the information, otherwise you're just trying to make the tests not fail when they should if some information is wrong.

File: interface/presenter/definition_presenter_test.go [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			presenter := NewDefinitionPresenter()
+			response := presenter.ResponseDefinitions(test.response)
+			definition := response.Definitions[0]
+
+			if test.parsedSuccessfull {
```
Comment (StradivariT) [inline interface/presenter/definition_presenter_test.go]: Please avoid redundant checks, assert.Equal will already take care of validating your tests values

File: interface/presenter/definition_presenter.go [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+func NewDefinitionPresenter() presenter.DefinitionPresenter {
+	return &definitionPresenter{}
+}
+
+// ResponseDefinitions return the list of definitions fulfilling the DefinitionPresenter interface
+func (l *definitionPresenter) ResponseDefinitions(definitionsList *model.List) *model.List {
+	for _, definition := range definitionsList.Definitions {
+		writtenParsed, err := time.Parse(UrbanLayout, definition.WrittenOn)
+		if err != nil {
+			continue
```
Comment (StradivariT) [inline interface/presenter/definition_presenter.go]: Handle the error properly, if something goes wrong then you should proably log the error and return with an error

File: interface/controller/definition_controller_test.go [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+	return args.Get(0).(*model.List), args.Error(1)
+}
+func (m MockInteractor) GetConcurrent() (*model.List, error) {
+	args := m.Called()
+	return args.Get(0).(*model.List), args.Error(1)
+}
+
+func TestGetDefinitions(t *testing.T) {
+	mockInteractor := MockInteractor{}
+	t.Run("success - valid request", func(t *testing.T) {
```
Comment (StradivariT) [inline interface/controller/definition_controller_test.go]: Be consistent with how you layout the tests througout the app, whether that's with the iterative format or with this individual tests format, I would suggest using the iterative format, it's easier to read

**Review by StradivariT — CHANGES_REQUESTED** [review]
